rest_framework/test.py

```diff
@@ -189,14 +189,9 @@ def _encode_data(self, data, format=None, content_type=None):
                     content_type, renderer.charset
                 )
 
-            if data is None:
-                ret = ''
-            else:
-                ret = renderer.render(data)
-
-            # Coerce text to bytes if required.
-            if isinstance(ret, str):
-                ret = ret.encode(renderer.charset)
+                # Coerce text to bytes if required.
+                if isinstance(ret, str):
+                    ret = ret.encode(renderer.charset)
```

Comment:
That used to be de-dented one level before #6511:

```suggestion
            # Coerce text to bytes if required.
            if isinstance(ret, str):
                ret = ret.encode(renderer.charset)
```

On the other hand, I don't see how `ret.encode(renderer.charset)` can work if `renderer.charset` is falsy (most likely `None`)... I'm _guessing_ the `if` branch at line 187 is always `True`.

All in all, I'm not sure my above suggestion matters, but if someone can think of a test case where it does, I'm open to hear it.

Reply:
I did this as charset can be None and encode raises an exception when None is passed. As far as I understand it, when a renderer sets charset to None (e.g. JSONRenderer) it handles the encoding on its own and returns bytes. However, if the renderer faultily returns str and has charset set to None, an exception occurred before this change.

My fix indenting the str check when no charset is set, actually avoids the exception which, when thinking about it again, might make things worse as it is a misconfiguration. One could argue an assert might be better in this case. I guess though if so that should be better part of another PR and I assume it has not been an issue in the past, so let's keep it as it was.

I will remove the indent again then.